Release data 26/03/2021

ESapenaVentura opened the pull request:
**Versions updated:**

system/links.json - v3.0.0
module/protocol/matrix.json - v1.0.0
type/protocol/analysis/analysis_protocol.json - v9.2.0
type/file/analysis_file.json - v6.3.0
core/file/file_core.json - v6.2.0
type/file/sequence_file.json - v9.3.0
type/file/supplementary_file.json - v2.3.0
type/file/reference_file.json - v3.3.0
type/file/image_file.json - v2.3.0
module/ontology/data_use_ontology.json - v1.0.0
type/project/project.json - v14.2.0

**Issues addressed**:
- https://github.com/HumanCellAtlas/metadata-schema/issues/1372
- https://github.com/HumanCellAtlas/metadata-schema/issues/1344
- https://github.com/HumanCellAtlas/metadata-schema/issues/1336
- https://github.com/HumanCellAtlas/metadata-schema/issues/1333

**New subgraphs?**:
- `tests/links/d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json`: Cell suspension --> Analysis files

aaclan-ebi commented:
> The path to the test data should not encode a date or time or any other version signifier. We want Git to do the version tracking for us. We can't release a new set of test data for every schema release. That would defeat the purpose of letting the Git history reflect how the metadata evolves. The schema evolution should be reflected in actual commit differences to existing files, not just in the accumulation of new metadata files, leaving in place the old files that comply with the old schema.

Hi @hannes-ucsc,  do you mean the version timestamp in the filenames? All timestamp values are using `2021-01-01T00:00:00.000000Z`. Instead of 0, we used a "zero" timestamp value which will never change. Is this a problem?

Also, there are 2 commits in the PR, 1st commit contains the base test data, the 2nd commit contains the diff which is what we want.

hannes-ucsc commented:
Sorry, I made a mistake. Disregard the my [previous comment](https://github.com/HumanCellAtlas/schema-test-data/pull/8#pullrequestreview-624616301). I took the branch name for a directory.

hannes-ucsc commented on tests/descriptors/analysis_file/cf5d9300-3b81-52eb-a02f-25fb1364419e_2021-01-01T00:00:00.000000Z.json:
```diff
@@ -0,0 +1,14 @@
+    "sha1": "4c7e22be4a2bdc4571d1a836cb1a1757d40363f1",
+    "sha256": "8908aec2bf49fdc557a5841352c52c8af857707c6f9c760ff51fb6e0858a6577",
+    "size": 143
+}
```
It would be nice to have a newline character at the end of the JSON so we don't always get this warning.

hannes-ucsc replied:
And by warning I mean this one:

![image](https://user-images.githubusercontent.com/5143256/113643219-a8d71380-9636-11eb-88cc-4334979fe535.png)

ESapenaVentura replied:
Thanks for the suggestion! We will modify the post-processor script to add the newline

aaclan-ebi replied:
json.dump doesn't add a trailing new line. We could add it, but just out of curiosity, is it a good practice to add a newline at the end of the file?

aaclan-ebi replied:
read a bit on this https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file , https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline/729795 it's very unlikely that it will cause us problems but yeah i guess we can just add it

ESapenaVentura replied:
@hannes-ucsc we have fixed the issue with the trailing newline, would you prefer for us to re-generate the data or would a third commit + squash work?

hannes-ucsc replied:
If this were the only issue with the PR then a 3rd commit would be fine. But there was also

> The second commit adds a links.json but also renames the links.json from the first commit. Why does it do that? Shouldn't the second commit simply add a links.json and leave the other one as is? I worry that we have a non-determinism in the UUID generation for the links.json files.

which looks like @aaclan-ebi already addressed but we should verify her fix works. Ideally, you would reset this branch to master and then run the tooling twice as (I assume) you did the first time around. With the fix for the links ID non-determinism we should not observe the second commit to touch the links.json added by the first commit, correct?

ESapenaVentura replied:
Yes, you are correct!

hannes-ucsc reviewed (changes requested):
Regarding the two commits: It seems that each commit adds a h5ad file. Not sure we need to structure the history that way. We want to be able to correlate a diff in this repository to a diff in the schema repository. So if a schema PR renames an property in a schema, there should be a one or two line diff in the schema PR and as many one line changes in the test data PR as there are instances of that schema in the test data repo.

Anyways, I think the two commits can be squashed simply because the test data project is made up exclusively of matrix files. There are no other sequence of analysis files. There is no initial commit that adds anything worth looking at.

The second commit adds a `links.json` but also renames the `links.json` from the first commit. Why does it do that? Shouldn't the second commit simply add a `links.json` and leave the other one as is? I worry that we have a non-determinism in the UUID generation for the `links.json` files.

ESapenaVentura commented:
We have corrected the problem with the links.json (It was an issue with the processes IDs being auto-generated each time) and re-deployed the test data to match the last changes we pushed to staging (rolling back the project version)\

About the files: I don't think I get the request. Comparing both commits shows the difference with the updated schemas (https://github.com/HumanCellAtlas/schema-test-data/compare/dd0553a98cd4d2eff216a8b9e567a04be85b3e75..7213226d250126a1ac327142663b7221d8506849). In the future, there will only be 1 commit that will compare against master and showcase the updates in the schema, but in this first iteration we also needed to add the baseline data.

Let me know if you have more concerns/queries!

hannes-ucsc commented:
> About the files: I don't think I get the request. Comparing both commits shows the difference with the updated schemas

Generally speaking, you can't force push a PR branch and then use the NEW history to question/contradict a claim that I make about the OLD history.

I am writing this after I wrote

https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r612640587

It looks like you already did exactly what I proposed?

hannes-ucsc commented:
If the answer is yes, then Daniel and I can re-review. He's made good progress on the code that reads this staging area:

https://github.com/DataBiosphere/hca-metadata-api/pull/42

ESapenaVentura commented:
> Generally speaking, you can't force push a PR branch and then use the NEW history to question/contradict a claim that I make about the OLD history.
> 
> I am writing this after I wrote
> 
> [#8 (comment)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r612640587)
> 
> It looks like you already did exactly what I proposed?

Sorry about the missunderstanding, I meant the comment about the h5ad files! 

About the other question, yes that's what we did! 

~~We have fixed the newlines but there's been a small problem squashing the commits. Apologies, I take full responsibility about this one, I squashed Alegria's fix of the script into the commit with the regenerated test data. If you could please ignore the `post_process.py` file, everything else should be in place as requested so you can re-review the test data.~~

We have solved the problem with the squash so now only the updates are showing

Many thanks!

hannes-ucsc commented:
> > Generally speaking, you can't force push a PR branch and then use the NEW history to question/contradict a claim that I make about the OLD history.
> > I am writing this after I wrote
> > [#8 (comment)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r612640587)
> > It looks like you already did exactly what I proposed?
> 
> Sorry about the missunderstanding, I meant the comment about the h5ad files!

Which comment? Maybe post a link to it?

dsotirho-ucsc commented on tests/descriptors/analysis_file/f0c8ac7c-1c4a-5c9c-9286-c8b16208f9ae_2021-01-01T00:00:00.000000Z.json:
```diff
@@ -0,0 +1,14 @@
+{
+    "content_type": "application/gzip; dcp-type=data",
+    "crc32c": "4dab8938",
+    "describedBy": "https://schema.staging.data.humancellatlas.org/system/2.0.0/file_descriptor",
```
The value of `describedBy` here fails against the `https://schema.staging.data.humancellatlas.org/system/2.0.0/file_descriptor` schema because the regex `schema(\\.[a-z]+)?\\.humancellatlas` does not allow for extra `.` (dot) characters between `schema.` and `.humancellatlas`.

Other schemas like `https://schema.staging.data.humancellatlas.org/type/biomaterial/13.3.0/cell_suspension` do not have this issue due to a different regex approach `schema.(.*?)humancellatlas`

ESapenaVentura replied:
Thanks for catching this @danielsotirhos! 

This schema was created for the DCP2 so I think this is the first time we ever created it in ingest staging, so we overlooked that, apologies!

We will update this! We are discussing the approaches and we have reduced it to 2 questions I will like to ask:
- Do we want to treat this as a bugfix (No version update, re-publish schema) or do we want to add it to go through the usual process and update the version? (We'll need to regenerate test data if that's the case)
- Do we want to include it in the existing schema release PR, or do we want to just move it through dev, staging and master as a hotfix?

@hannes-ucsc @HumanCellAtlas/wranglers @aaclan-ebi

hannes-ucsc replied:
> Do we want to treat this as a bugfix (No version update, re-publish schema) or do we want to add it to go through the usual process and update the version? (We'll need to regenerate test data if that's the case)

Since this hasn't been released yet, I think it's OK to not bump the version. I don't know what repercussions this has for deploying up to staging. OTOH, it wouldn't affect us if the version were bumped, either. Re-generating the test metadata for this PR would be an interesting test of the idempotency of that process provided that you don't force push after regenerating.

> Do we want to include it in the existing schema release PR, or do we want to just move it through dev, staging and master as a hotfix?

Since this is necessary to get organic CGMs working, it should be included in the respective CGM PR against the schema repo.

aaclan-ebi replied:
I agree. We should just add the schema regex fix as part of the metadata-schema PR https://github.com/HumanCellAtlas/metadata-schema/pull/1371/files and regenerate the test data.

hannes-ucsc replied:
@danielsotirhos please verify resolution.

dsotirho-ucsc replied:
Confirmed the issue with the `file_descriptor` schema has been resolved, and that all files in this canned staging area as of commit 2c7fbd0e0b7804abeb21ddb0ae89724950218257 pass JSON validation against their schema.

dsotirho-ucsc commented on tests/links/37e91a9f-b04e-5313-bbc0-4c394406247e_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json:
```diff
@@ -0,0 +1,80 @@
+                    "output_type": "specimen_from_organism"
+                }
+            ],
+            "process_id": "37e91a9f-b04e-5313-bbc0-4c394406247e",
```
@hannes-ucsc

hannes-ucsc replied:
Can't use the same UUID for two different entities (links and process). If you want determinism, you can derive one UUID from the other using the v5 UUID algo.

aaclan-ebi replied:
@hannes-ucsc the links.json reuses the assay process UUID for that subgraph by design in Ingest. Will this cause any problems downstream?

aaclan-ebi replied:
This has been happening in production as well ever since DCP2 MVP release.

hannes-ucsc replied:
The general assumption with entity IDs (especially UUID) is that they uniquely identify an entity. It would have been better to send that through a UUID v5 hash. So `links_id := UUIDv5(process_id)` instead of `links_id := process_id`.

> This has been happening in production as well ever since DCP2 MVP release.

I just confirmed with BQ that this is the case. It doesn't cause any issues with Azul, but that is only by coincidence, I could have easily written code that puts entities and links in a dictionary keyed by their ID.

Since what I objected to is actual production behavior, we can leave it as is. But we need to document this quirk:

https://github.com/HumanCellAtlas/dcp2/issues/20

ESapenaVentura commented:
> > > Generally speaking, you can't force push a PR branch and then use the NEW history to question/contradict a claim that I make about the OLD history.
> > > I am writing this after I wrote
> > > [#8 (comment)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r612640587)
> > > It looks like you already did exactly what I proposed?
> > 
> > 
> > Sorry about the missunderstanding, I meant the comment about the h5ad files!
> 
> Which comment? Maybe post a link to it?

This comment https://github.com/HumanCellAtlas/schema-test-data/pull/8#pullrequestreview-628277075

dsotirho-ucsc commented on tests/links/37e91a9f-b04e-5313-bbc0-4c394406247e_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json:
```diff
@@ -0,0 +1,80 @@
+                    "output_type": "analysis_file"
+                }
+            ],
+            "process_id": "d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7",
```
A metadata file is missing from `schema-test-data/tests/metadata/process/` for this process ID.

hannes-ucsc replied:
If this is resolved, please click

![image](https://user-images.githubusercontent.com/5143256/116593084-923e7800-a8d5-11eb-806b-2d1478211b6a.png)

dsotirho-ucsc commented on tests/links/37e91a9f-b04e-5313-bbc0-4c394406247e_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json:
```diff
@@ -0,0 +1,80 @@
+                    "output_type": "cell_suspension"
+                }
+            ],
+            "process_id": "a3843530-046c-5ec0-bbf4-a4f0929b60af",
```
A metadata file is missing from `schema-test-data/tests/metadata/process/` for this process ID.

hannes-ucsc replied:
Same here

dsotirho-ucsc commented on tests/links/37e91a9f-b04e-5313-bbc0-4c394406247e_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json:
```diff
@@ -0,0 +1,80 @@
+                    "output_type": "specimen_from_organism"
+                }
+            ],
+            "process_id": "37e91a9f-b04e-5313-bbc0-4c394406247e",
```
A metadata file is missing from `schema-test-data/tests/metadata/process/` for this process ID.

hannes-ucsc replied:
Same here

danielsotirhos reviewed (changes requested):
The file `tests/links/37e91a9f-b04e-5313-bbc0-4c394406247e_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json` lists `process_id` for files that appear to be missing from  `schema-test-data/tests/metadata/process/`

hannes-ucsc commented:
> > > > Generally speaking, you can't force push a PR branch and then use the NEW history to question/contradict a claim that I make about the OLD history.
> > > > I am writing this after I wrote
> > > > [#8 (comment)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r612640587)
> > > > It looks like you already did exactly what I proposed?
> > > 
> > > 
> > > Sorry about the missunderstanding, I meant the comment about the h5ad files!
> > 
> > 
> > Which comment? Maybe post a link to it?
> 
> This comment [#8 (review)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#pullrequestreview-628277075)

That comment does not mention h5ad files.

ESapenaVentura commented:
> > > > > Generally speaking, you can't force push a PR branch and then use the NEW history to question/contradict a claim that I make about the OLD history.
> > > > > I am writing this after I wrote
> > > > > [#8 (comment)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r612640587)
> > > > > It looks like you already did exactly what I proposed?
> > > > 
> > > > 
> > > > Sorry about the missunderstanding, I meant the comment about the h5ad files!
> > > 
> > > 
> > > Which comment? Maybe post a link to it?
> > 
> > 
> > This comment [#8 (review)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#pullrequestreview-628277075)
> 
> That comment does not mention h5ad files.

It does, in the very first line

aaclan-ebi commented:
@danielsotirhos I am not sure why there were missing metadata files and incorrect links.json filenames. But I've just re-run the post-processor against the original test data and I believe that corrected the issue.

ESapenaVentura commented:
@danielsotirhos as alegria pointed out, she has added the missing files and we have corrected the regex.

It should be fine now! Please let us know if you find any further issues

hannes-ucsc reviewed (changes requested):
https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r616875385

wasn't addressed. The UUID of two processes is reused as that of links files:

```
$ find . | grep -E '4da04038-adab-59a9-b6c4-3a61242cc972|d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7'
./tests/metadata/process/4da04038-adab-59a9-b6c4-3a61242cc972_2021-01-01T00:00:00.000000Z.json
./tests/metadata/process/d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7_2021-01-01T00:00:00.000000Z.json
./tests/links/4da04038-adab-59a9-b6c4-3a61242cc972_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json
./tests/links/d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json
```

Each entity must have a universally unique ID (hence the name UUID), even if they are of different types.

hannes-ucsc commented:
> > > > > > Generally speaking, you can't force push a PR branch and then use the NEW history to question/contradict a claim that I make about the OLD history.
> > > > > > I am writing this after I wrote
> > > > > > [#8 (comment)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r612640587)
> > > > > > It looks like you already did exactly what I proposed?
> > > > > 
> > > > > 
> > > > > Sorry about the missunderstanding, I meant the comment about the h5ad files!
> > > > 
> > > > 
> > > > Which comment? Maybe post a link to it?
> > > 
> > > 
> > > This comment [#8 (review)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#pullrequestreview-628277075)
> > 
> > 
> > That comment does not mention h5ad files.
> 
> It does, in the very first line

Ahh, I stand corrected. That comment was more about the history than the actual files, hence my confusion with your reference to my comment as being about h5ad files. But, technically you are correct.

aaclan-ebi commented:
> [#8 (comment)](https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r616875385)
> 
> wasn't addressed. The UUID of two processes is reused as that of links files:
> 
> ```
> $ find . | grep -E '4da04038-adab-59a9-b6c4-3a61242cc972|d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7'
> ./tests/metadata/process/4da04038-adab-59a9-b6c4-3a61242cc972_2021-01-01T00:00:00.000000Z.json
> ./tests/metadata/process/d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7_2021-01-01T00:00:00.000000Z.json
> ./tests/links/4da04038-adab-59a9-b6c4-3a61242cc972_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json
> ./tests/links/d7b8cbff-aee9-5a05-a4a1-d8f4e720aee7_2021-01-01T00:00:00.000000Z_90bf705c-d891-5ce2-aa54-094488b445c6.json
> ```
> 
> Each entity must have a universally unique ID (hence the name UUID), even if they are of different types.

Replied here https://github.com/HumanCellAtlas/schema-test-data/pull/8#discussion_r623140955

ruchim reviewed (approved):
haven't done a deep dive, but approving for fullness.

hannes-ucsc reviewed (approved):
~@danielsotirhos could you also please verify whether the reused~

aherbst-broad reviewed (approved):
Looks good.
